### PR TITLE
TagBox - Provide the capability to specify the event that should trigger adding of a custom item

### PR DESCRIPTION
**Pull request by marker-dao**

### CODE REVIEW

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js [hunk 1]
```diff
@@ -3747,19 +3747,19 @@ QUnit.module('searchEnabled', moduleSetup, () => {
 
     QUnit.test('search was work if acceptCustomValue is set to true', function(assert) {
         const $element = $('#tagBox').dxTagBox({
-            dataSource: ['item 1', 'element 1', 'item 2'],
+            dataSource: ['item 1', 'element 1', 'item 2', 'item 3'],
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js]: Why should we change this test?

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js [hunk 2]
```diff
@@ -5071,6 +5071,41 @@ QUnit.module('the \'fieldTemplate\' option', moduleSetup, () => {
     });
 });
 
+
+QUnit.module('the \'valueChangeEvent\' option', moduleSetup, () => {
+    ['keyup', 'blur', 'change', 'input', 'focusout'].forEach((eventName) => {
+        QUnit.test(`custom item has been added when valueChangeEvent='${eventName}'`, function(assert) {
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js]: Shouldn't we also test combinations of the events? Won't item be added 2 times if we specify `valueChangeEvent="change focusout"` or like this?

Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js]: Shouldn't we also test that custom item is added to a datasource? (how many times custom item creating event was called)

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js (hunk 1 above)
Comment (marker-dao) [reply testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js]: https://github.com/DevExpress/DevExtreme/pull/22772/commits/87a7c60cf7177115eeeabf74aeb88d3ca9866bd2

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js (hunk 2 above)
Comment (marker-dao) [reply testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js]: https://github.com/DevExpress/DevExtreme/pull/22772/commits/14e8e6a966706fbe493510d28654286a4bdfd2ad

Comment (marker-dao) [reply testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js]: https://github.com/DevExpress/DevExtreme/pull/22772/commits/14e8e6a966706fbe493510d28654286a4bdfd2ad

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js [hunk 3]
```diff
@@ -5071,6 +5071,75 @@ QUnit.module('the \'fieldTemplate\' option', moduleSetup, () => {
     });
 });
 
+
+QUnit.module('the \'valueChangeEvent\' option', moduleSetup, () => {
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js]: To simplify search when u copy test name from the browser page
```suggestion
QUnit.module('the "valueChangeEvent" option', moduleSetup, () => {
```

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js [hunk 4]
```diff
@@ -5071,6 +5071,75 @@ QUnit.module('the \'fieldTemplate\' option', moduleSetup, () => {
...
+                    },
+                });
+
+                const $input = $tagBox.find(`.${TEXTBOX_CLASS}`);
+                const keyboard = keyboardMock($input);
+                const customValue = 't';
+                // const [first, second] = eventValue.split(' ');
+                const eventsArray = eventValue.split(' ');
+
+                switch(eventValue) {
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js]: Let's implement separate tests to simplify debug and future support. We can move initialization to `beforeEach` to reduce copypaste.

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js [hunk 5]
```diff
@@ -5071,6 +5071,75 @@ QUnit.module('the \'fieldTemplate\' option', moduleSetup, () => {
...
+                        $($input).trigger(eventValue);
+                        break;
+                    default:
+                        keyboard.type(customValue);
+                        $($input).trigger(eventsArray[0]);
+                        $($input).trigger(eventsArray[1]);
+                        break;
+                }
+
+                assert.strictEqual(onCustomItemCreatingSpy.callCount, 1, 'the \'onCustomItemCreating\' was fired once');
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js]:
```suggestion
                assert.strictEqual(onCustomItemCreatingSpy.callCount, 1, 'the "onCustomItemCreating" was fired once');
```

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js (hunk 3 above)
Comment (marker-dao) [reply testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js]: https://github.com/DevExpress/DevExtreme/pull/22772/commits/42f158af2eaaba57402677771dc9b9eb3bac416d

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js (hunk 5 above)
Comment (marker-dao) [reply testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js]: https://github.com/DevExpress/DevExtreme/pull/22772/commits/42f158af2eaaba57402677771dc9b9eb3bac416d

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js (hunk 4 above)
Comment (marker-dao) [reply testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js]: https://github.com/DevExpress/DevExtreme/pull/22772/commits/d60c77cdb5d8d06e7b00552fa2aaa61a62a1a3f0

File: js/ui/select_box.js [hunk 6]
```diff
@@ -714,6 +714,12 @@ const SelectBox = DropDownList.inherit({
 
     _customItemAddedHandler: function(e) {
         const searchValue = this._searchValue();
+
+        if(!searchValue) {
```
Comment (ksercs) [inline js/ui/select_box.js]: Why do we write it? Where is an appropriate test?

Comment (marker-dao) [reply js/ui/select_box.js]: https://github.com/DevExpress/DevExtreme/pull/22772/commits/33a7265c92cf6e71f01d918a990476978643de40

File: js/ui/select_box.js [hunk 7]
```diff
@@ -596,6 +605,10 @@ const SelectBox = DropDownList.inherit({
         return this.callBase() && $(activeElement).closest(this._input()).length > 0;
     },
 
+    _getChangeValueEventOptionName: function() {
```
Comment (ksercs) [inline js/ui/select_box.js]:
```suggestion
    _getValueChangeEventOptionName: function() {
```

File: js/ui/text_box/ui.text_editor.base.js [hunk 8]
```diff
@@ -631,8 +635,14 @@ const TextEditorBase = Editor.inherit({
...
 
+    _changeValueEventIncludesBlur: function() {
+        const eventOption = this.option(this._getChangeValueEventOptionName());
+
+        return eventOption.includes('blur');
+    },
+
     _preventNestedFocusEvent: function(event) {
-        if(event.isDefaultPrevented()) {
+        if(event.isDefaultPrevented() || this._changeValueEventIncludesBlur()) {
```
Comment (ksercs) [inline js/ui/text_box/ui.text_editor.base.js]: If you write changed on text_editor level, you should add tests for this change in text_editor. Are u sure that this changes are necessary exactly here?

File: testing/tests/DevExpress.ui.widgets.editors/selectBox.tests.js [hunk 9]
```diff
@@ -69,6 +69,8 @@ const KEY_SPACE = ' ';
 
 const TIME_TO_WAIT = 500;
 
+const CHANGE_VALUE_EVENT_OPTIONS = ['valueChangeEvent', 'customItemCreateEvent'];
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/selectBox.tests.js]:
```suggestion
const VALUE_CHANGE_EVENT_OPTIONS = ['valueChangeEvent', 'customItemCreateEvent'];
```

File: js/ui/select_box.js (hunk 7 above)
Comment (marker-dao) [reply js/ui/select_box.js]: https://github.com/DevExpress/DevExtreme/pull/22772/commits/f54730c001335b6e49b056d0a5a677bc93bd16bc

File: testing/tests/DevExpress.ui.widgets.editors/selectBox.tests.js (hunk 9 above)
Comment (marker-dao) [reply testing/tests/DevExpress.ui.widgets.editors/selectBox.tests.js]: https://github.com/DevExpress/DevExtreme/pull/22772/commits/f54730c001335b6e49b056d0a5a677bc93bd16bc

Comment (marker-dao) [reply testing/tests/DevExpress.ui.widgets.editors/selectBox.tests.js]: https://github.com/DevExpress/DevExtreme/pull/22772/commits/f54730c001335b6e49b056d0a5a677bc93bd16bc

File: js/ui/text_box/ui.text_editor.base.js (hunk 8 above)
Comment (marker-dao) [reply js/ui/text_box/ui.text_editor.base.js]: https://github.com/DevExpress/DevExtreme/pull/22772/commits/f81dc1c6a5a31f5b9a0bb82c18ba0b68db2016e5
https://github.com/DevExpress/DevExtreme/pull/22772/commits/aca844e4b81e0c308de49409c6b672d9a57ccfee

File: js/ui/select_box.js [hunk 10]
```diff
@@ -529,8 +538,18 @@ const SelectBox = DropDownList.inherit({
         }).bind(this));
     },
 
+    _changeValueEventIncludesBlur: function() {
```
Comment (ksercs) [inline js/ui/select_box.js]:
```suggestion
    _valueChangeEventIncludesBlur: function() {
```

File: js/ui/select_box.js [hunk 11]
```diff
@@ -529,8 +538,18 @@ const SelectBox = DropDownList.inherit({
...
+        const eventOption = this.option(this._getValueChangeEventOptionName());
```
Comment (ksercs) [inline js/ui/select_box.js]:
```suggestion
        const valueChangeEvent = this.option(this._getValueChangeEventOptionName());
```

File: js/ui/text_box/ui.text_editor.base.js [hunk 12]
```diff
@@ -746,6 +750,7 @@ const TextEditorBase = Editor.inherit({
 
         switch(name) {
             case 'valueChangeEvent':
+            case 'customItemCreateEvent':
```
Comment (ksercs) [inline js/ui/text_box/ui.text_editor.base.js]: there is now such option on text editor => we should not handle its change here

**Review by IlyaKhD — DISMISSED** [review]

File: js/ui/select_box.js (hunk 10 above)
Comment (marker-dao) [reply js/ui/select_box.js]: https://github.com/DevExpress/DevExtreme/pull/22772/commits/4c59742555c2e9e4b45320926142f9f53fee8cd8

File: js/ui/select_box.js (hunk 11 above)
Comment (marker-dao) [reply js/ui/select_box.js]: https://github.com/DevExpress/DevExtreme/pull/22772/commits/4c59742555c2e9e4b45320926142f9f53fee8cd8

File: js/ui/text_box/ui.text_editor.base.js (hunk 12 above)
Comment (marker-dao) [reply js/ui/text_box/ui.text_editor.base.js]: https://github.com/DevExpress/DevExtreme/pull/22772/commits/4c59742555c2e9e4b45320926142f9f53fee8cd8

**Review by ksercs — APPROVED** [review]

**Review by EugeniyKiyashko — APPROVED** [review]

**Pull request merged**